core/src/main/java/co/aikar/commands/BaseCommand.java

```diff
@@ -515,39 +767,62 @@ public boolean canExecute(CommandIssuer issuer, RegisteredCommand<?> cmd) {
         }
     }
 
+    /**
+     * Gets all subcommands which are possible to tabcomplete.
+     *
+     * @param issuer
+     *         The command issuer.
+     * @param args
+     *
+     * @return
+     */
     List<String> getCommandsForCompletion(CommandIssuer issuer, String[] args) {
+        if (!hasPermission(issuer)) {
+            return ImmutableList.of();
+        }
         final Set<String> cmds = new HashSet<>();
         final int cmdIndex = Math.max(0, args.length - 1);
         String argString = ApacheCommonsLangUtil.join(args, " ").toLowerCase();
+
         for (Map.Entry<String, RegisteredCommand> entry : subCommands.entries()) {
             final String key = entry.getKey();
             if (key.startsWith(argString) && !CATCHUNKNOWN.equals(key) && !DEFAULT.equals(key)) {
                 final RegisteredCommand value = entry.getValue();
-                if (!value.hasPermission(issuer)) {
+                if (!canExecute(issuer, value)) {
```

Comment:
while it seems ok in principle, this is a breaking change as one expects permission checks to already be done for canExecute, and anyone overriding it may of not implemented a permission check (This applies for the above revert request too)